MathLib/mechanics_invariants.h

```diff
@@ -0,0 +1,189 @@
+
+    /// Von Mises equivalent stress.
+    /// \note The input vector must have trace equal zero.
+    static double equivalentStress(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_v)
+    {
+        static_assert(KelvinVectorSize == 4 || KelvinVectorSize == 6,
+                      "KelvinVector invariants, equivalentStress: input vector "
+                      "of size different than 4 or 6 is not allowed.");
+        assert(std::abs(trace(deviatoric_v)) <
+               std::numeric_limits<double>::epsilon());
+        return std::sqrt(3 * J2(deviatoric_v));
+    }
+
+    /// Second invariant of deviatoric tensor.
+    /// \note The input vector must have trace equal zero.
+    static double J2(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_v)
+    {
+        static_assert(KelvinVectorSize == 4 || KelvinVectorSize == 6,
+                      "KelvinVector invariants, J2: input vector of size "
+                      "different than 4 or 6 is not allowed.");
+        assert(std::abs(trace(deviatoric_v)) <
+               std::numeric_limits<double>::epsilon());
+        return 0.5 * deviatoric_v.transpose() * deviatoric_v;
+    }
+
+    /// Third invariant, equal to determinant of a deviatoric tensor.
+    /// \note The input vector must have trace equal zero.
+    static double J3(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_v)
+    {
+        static_assert(KelvinVectorSize == 4 || KelvinVectorSize == 6,
+                      "KelvinVector invariants, J3: input vector of size "
+                      "different than 4 or 6 is not allowed.");
+        assert(std::abs(trace(deviatoric_v)) <
+               std::numeric_limits<double>::epsilon());
+        return determinant(deviatoric_v);
+    }
+
+    /// Trace of the corresponding tensor.
+    static double trace(Eigen::Matrix<double, KelvinVectorSize, 1> const& v)
+    {
+        static_assert(KelvinVectorSize == 4 || KelvinVectorSize == 6,
+                      "KelvinVector invariants, trace: input vector of size "
+                      "different than 4 or 6 is not allowed.");
+        return v.template topLeftCorner<3, 1>().sum();
+    }
+};
+
+template <>
+inline double Invariants<6>::determinant(Eigen::Matrix<double, 6, 1> const& v)
+{
+    return v(0) * v(1) * v(2) + v(3) * v(4) * v(5) / std::sqrt(2.) -
+           v(3) * v(3) * v(2) / 2. - v(4) * v(4) * v(0) / 2. -
+           v(5) * v(5) * v(1) / 2.;
+}
+
+template <>
+inline double Invariants<4>::determinant(Eigen::Matrix<double, 4, 1> const& v)
+{
+    return v(2) * (v(0) * v(1) - v(3) * v(3) / 2.);
+}
+
+template <>
+inline Eigen::Matrix<double, 4, 1> Invariants<4>::inverse(
+    Eigen::Matrix<double, 4, 1> const& v)
+{
+    Eigen::Matrix<double, 4, 1> inv;
+    inv(0) = v(1) * v(2);
+    inv(1) = v(0) * v(2);
+    inv(2) = v(0) * v(1) - v(3) * v(3) / 2.;
+    inv(3) = - v(3) * v(2);
+    return inv / determinant(v);
```

Comment:
i am not sure if ROV works with this expression. `inv /= determinant(v); return inv;` might be better.